### ISSUE TITLE
Add a function equivalent to xsl:number

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#1427]
I propose adding a function to perform node numbering in a manner analogous to xsl:number (but without the formatting aspects, which can be handled using format-integer, and without multi-level numbering).

I envisage a function along the following lines:

```
node-number($node as element(), 
                         $from as (document-node()|element(), 
                         $count as fn($node as element()) as xs:boolean?)
   as xs:integer
```

The function returns the number of element nodes that satisfy all the following conditions:

* they are descendants of $from
* they are preceding-or-self nodes of $node
* they satisfy the $count predicate

$node defaults to the context node. $from defaults to the parent of $node. $count defaults to a function that returns true for an element that has the same name as $node, false otherwise.

With no arguments, node-number() applied to (say) a `<p>` element returns the number of preceding-sibling `<p>` elements plus one.

The rationale for this proposal is (a) to make the core functionality of xsl:number available in environments other than XSLT, and (b) within XSLT, to make it available in contexts such as a predicate of a match pattern where it is currently difficult or impossible to invoke `xsl:number` except by wrapping it in a user-defined variable or function.

**Comment by ChristianGruen** [comment]
If the function had no `$from` parameter, would the following call be more or less equivalent?

```xquery
node-number($node[ancestor::node()[. is $from]], $count)
```

Maybe I wouldn’t ask if I had already worked with `xsl:number`.

**Comment by michaelhkay** [comment]
>would the following call be more or less equivalent?

No, I don't think so. If you want to number equations within a chapter, you need something like

`node-number(., ancestor::chapter)`

to determine how many previous equations there are in the same chapter.

**Comment by Arithmeticus** [comment]
I like this idea. For some of the reasons that motivate this suggestion, I have generally avoided `xsl:number`, and preferred XPath.

If a zero-arity function is envisioned, then question marks should be added to the parameters.

I'd be curious to see what others thing, but for the default of `$from` I would prefer `ancestor::*[last()]`. Basically for the reason expressed in the italics: People could be easily lulled into thinking this gets the count of preceding sibling nodes. If you're getting preceding nodes, not preceding sibling nodes, then might as well get the whole gamut of preceding nodes in the context tree.
